### PR TITLE
Rework generated code for return+out in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR updates the generated code for return+out (when both return and out are non-optional) to be encoding-dependent:
- with the 1.1 encoding, the return is marshaled after the out parameters
- with other encodings (typically 2.0), the return is marshaled before the out parameters

### CODE REVIEW

File: cpp/src/slice2cs/Gen.cpp [hunk 1]
```diff
@@ -157,110 +157,191 @@ Slice::CsVisitor::~CsVisitor()
...
+    const string stream = "ostr";
+    string ns = getNamespace(operation->interface());
+
+    MemberList members = returnType ? operation->returnValues() : operation->parameters();
+    auto [requiredMembers, taggedMembers] = getSortedMembers(members);
 
     int bitSequenceIndex = -1;
     size_t bitSequenceSize = 0;
-    if (requiredParams.size() > 0)
+    if (requiredMembers.size() > 0)
```
Comment (pepone) [inline cpp/src/slice2cs/Gen.cpp:170]: we can put the code below in a single line

```
bitSequenceSize = returnType ? operation->returnBitSequenceSize() : operation->paramsBitSequenceSize();
```

File: cpp/src/slice2cs/Gen.cpp [hunk 2]
```diff
@@ -157,110 +157,191 @@ Slice::CsVisitor::~CsVisitor()
...
+        _out << nl << "if (" << stream << ".Encoding != ZeroC.Ice.Encoding.V1_1)";
+        _out << sb;
+    }
 
+    bool repeat = false;
+    do
+    {
         if (bitSequenceSize > 0)
         {
             _out << nl << "var bitSequence = " << stream << ".WriteBitSequence(" << bitSequenceSize << ");";
```
Comment (pepone) [inline cpp/src/slice2cs/Gen.cpp:195]: We are now declaring the `bitSequence`, for the if and the else blocks, we can move it out the if/else and just reset bitSequenceIndex when we repeat

File: cpp/src/slice2cs/Gen.cpp [hunk 3]
```diff
@@ -157,110 +157,191 @@ Slice::CsVisitor::~CsVisitor()
...
+        _out << sb;
+    }
+
+    bool repeat = false;
+
+    do
+    {
         if (bitSequenceSize > 0)
         {
             _out << nl << "var bitSequence = " << stream << ".ReadBitSequence(" << bitSequenceSize << ");";
```
Comment (pepone) [inline cpp/src/slice2cs/Gen.cpp:288]: same here we can avoid declaring bitSequence twice

File: cpp/src/slice2cs/Gen.cpp [hunk 4]
```diff
@@ -157,110 +157,191 @@ Slice::CsVisitor::~CsVisitor()
...
+            _out << sb;
+
+            // Repeat after updating requiredMembers
+            repeat = true;
+            requiredMembers.push_back(requiredMembers.front());
+            requiredMembers.pop_front();
+        }
     }
+    while (repeat)
+        ;
```
Comment (pepone) [inline cpp/src/slice2cs/Gen.cpp:242]:
```suggestion
    while (repeat);
```

**Review by pepone — APPROVED** [review]
Looks good, just add a few comments

File: cpp/src/slice2cs/Gen.cpp [hunk 5]
```diff
@@ -157,110 +157,191 @@ Slice::CsVisitor::~CsVisitor()
...
-    const string stream = customStream.empty() ? "istr" : customStream;
-    string ns = getNamespace(op->interface());
+    const string stream = "istr";
+    string ns = getNamespace(operation->interface());
+
+    MemberList members = returnType ? operation->returnValues() : operation->parameters();
+    auto [requiredMembers, taggedMembers] = getSortedMembers(members);
 
     int bitSequenceIndex = -1;
     size_t bitSequenceSize = 0;
```
Comment (InsertCreativityHere) [inline cpp/src/slice2cs/Gen.cpp:260]: This could also be reduced to a single size.

File: cpp/src/slice2cs/Gen.cpp (hunk 2 above)
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:195]: I think it's clearer to do the whole marshaling/unmarshaling in each block, including these redundant bitSequence variables. I initially tried to create a block just for the return value, but it gets more complicated (and actually didn't work).

File: cpp/src/slice2cs/Gen.cpp (hunk 1 above)
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:170]: Fixed as suggested.

File: cpp/src/slice2cs/CsUtil.h [hunk 6]
```diff
@@ -44,8 +44,8 @@ std::string helperName(const TypePtr&, const std::string&);
 
 std::string builtinSuffix(const BuiltinPtr&);
 
-std::string resultType(const OperationPtr&, const std::string&, bool);
-std::string resultTask(const OperationPtr&, const std::string&, bool);
+std::string returnTypeStr(const OperationPtr&, const std::string& ns, bool dispatch);
+std::string returnTaskStr(const OperationPtr&, const std::string& ns, bool dispatch);
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/slice2cs/CsUtil.h:48]: I agree, fixed.

File: cpp/src/slice2cs/Gen.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:242]: Fixed.

File: cpp/src/slice2cs/Gen.cpp (hunk 2 above)
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:195]: Fixed as suggested.

File: cpp/src/slice2cs/Gen.cpp (hunk 3 above)
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:288]: Fixed a suggested.

File: cpp/src/slice2cs/Gen.cpp (hunk 4 above)
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:242]: Fixed.

File: cpp/src/slice2cs/Gen.cpp [hunk 7]
```diff
@@ -157,110 +157,191 @@ Slice::CsVisitor::~CsVisitor()
...
+
+    bool write11ReturnLast = returnType && operation->hasReturnAndOut() && !members.front()->tagged() &&
+        requiredMembers.size() > 1;
+    if (write11ReturnLast)
+    {
+        _out << nl << "if (" << stream << ".Encoding != ZeroC.Ice.Encoding.V1_1)";
+        _out << sb;
+    }
 
+    bool repeat = false;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:190]: Fixed as suggested.

File: cpp/src/slice2cs/Gen.cpp [hunk 8]
```diff
@@ -157,110 +157,191 @@ Slice::CsVisitor::~CsVisitor()
...
+    bool read11ReturnLast = returnType && operation->hasReturnAndOut() && requiredMembers.size() > 1 &&
+        !members.front()->tagged();
+
+    if (read11ReturnLast)
+    {
+        _out << nl << "if (" << stream << ".Encoding != ZeroC.Ice.Encoding.V1_1)";
+        _out << sb;
+    }
+
+    bool repeat = false;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:282]: Same fix.

File: cpp/src/slice2cs/Gen.cpp [hunk 9]
```diff
@@ -157,110 +157,191 @@ Slice::CsVisitor::~CsVisitor()
...
-    const string stream = customStream.empty() ? "ostr" : customStream;
-    string ns = getNamespace(op->interface());
+    const string stream = "ostr";
+    string ns = getNamespace(operation->interface());
+
+    MemberList members = returnType ? operation->returnValues() : operation->parameters();
+    auto [requiredMembers, taggedMembers] = getSortedMembers(members);
 
     int bitSequenceIndex = -1;
     size_t bitSequenceSize = 0;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:169]: Fixed.

File: cpp/src/slice2cs/Gen.cpp (hunk 5 above)
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:260]: Fixed.

**Pull request merged**